security/tests/integration/cmd/main.go

```diff
@@ -0,0 +1,276 @@
+			!bytes.Equal(initialSecret.Data[controller.CertChainID], secret.Data[controller.CertChainID]) {
+			glog.Infof("certificates were successfully rotated")
+			return nil
+		}
+	}
+	return fmt.Errorf("failed to validate certificate rotation")
+}
+
+func readFile(path string) (string, error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return "", err
+	}
+	return string(data), nil
+}
+
+func readURI(uri string) (string, error) {
+	resp, err := http.Get(uri)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", nil
+	}
+
+	return string(bodyBytes), nil
+}
+
+func runNodeAgentCertificateTests(env *integration.NodeAgentTestEnv, rootCert string, certChain string) error {
+	orgRootCert, err := readFile(rootCert)
+	if err != nil {
+		return fmt.Errorf("unable to read original root certificate: %v", rootCert)
+	}
+
+	orgCertChain, err := readFile(certChain)
+	if err != nil {
+		return fmt.Errorf("unable to read original certificate chain: %v", certChain)
+	}
+
+	nodeAgentIPAddress, err := env.GetNodeAgentIPAddress()
+	if err != nil {
+		return fmt.Errorf("external IP address of NodeAgent is not ready")
+	}
+
+	term := certValidationInterval
+	for i := 0; i < certValidateRetry; i++ {
+		if i > 0 {
+			glog.Infof("retry checking certificate update and validation in %v seconds", term)
+			time.Sleep(time.Duration(term) * time.Second)
+			term = term * 2
+		}
+
+		certPEM, err := readURI(fmt.Sprintf("http://%v:8080/cert", nodeAgentIPAddress))
+		if err != nil {
+			glog.Errorf("failed to read the certificate of NodeAgent: %v", err)
+			continue
+		}
+
+		rootPEM, err := readURI(fmt.Sprintf("http://%v:8080/root", nodeAgentIPAddress))
+		if err != nil {
+			glog.Errorf("failed to read the root certificate of NodeAgent: %v", err)
+			continue
+		}
+
+		if orgRootCert != rootPEM {
+			return fmt.Errorf("invalid root certificate was downloaded")
+		}
+
+		if orgCertChain == certPEM {
+			glog.Error("certificate chain was not updated yet")
+			continue
+		}
+
+		roots := x509.NewCertPool()
+		ok := roots.AppendCertsFromPEM([]byte(orgRootCert))
+		if !ok {
+			return fmt.Errorf("failed to parse root certificate")
+		}
+
+		block, _ := pem.Decode([]byte(certPEM))
+		if block == nil {
+			return fmt.Errorf("failed to parse certificate PEM")
+		}
+
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return fmt.Errorf("failed to parse certificate: %v", err)
+		}
+
+		opts := x509.VerifyOptions{
+			Roots: roots,
+		}
+
+		if _, err := cert.Verify(opts); err != nil {
+			return fmt.Errorf("failed to verify certificate: %v", err)
+		}
+
+		return nil
+	}
+
+	return fmt.Errorf("failed to check certificate update and validate after %v retry", certValidateRetry)
+}
+
+func main() {
+	kubeconfig := flag.String("kube-config", "", "path to kubeconfig file")
+	hub := flag.String("hub", "", "Docker hub that the Istio CA image is hosted")
```

Comment:
For me, it makes more sense to put flag `hub` and `tag` down to component level

Reply:
Thoughts?

Reply:
main.go is deprecated.

Reply:
Right, but we still keep these flags in testMain(). I think it makes more sense logically to move `hub` and `tag` to component level and kubeconfig can be moved to env or so :)